### ISSUE TITLE
Incorrect "Nullability doesn't match the target delegate" when tuple element names don't explicitly match

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#74493]
**Version Used**: 17.11.0 Preview 3.0

The following code erroneously causes this warning to appear for each lambda in the Error List and at the command line, though squiggles are not shown:

> ⚠️ CS8621 Nullability of reference types in return type of 'lambda expression' doesn't match the target delegate 'Func<(string, object?), ?>' (possibly because of nullability attributes).

```cs
class C
{
    void M((string Name, object? Value)[] parameters)
    {
        _ = parameters.Append(("A", "A")).ToDictionary(
            tuple => tuple.Name,
            tuple => tuple.Value);
    }
}
```

If you name the elements in the tuple expression to exactly match the tuple element names in the collection, the warning goes away.

```cs
class C
{
    void M((string Name, object? Value)[] parameters)
    {
        _ = parameters.Append((Name: "A", Value: "A")).ToDictionary(
            tuple => tuple.Name,
            tuple => tuple.Value);
    }
}
```

For each element where a name is missing or does not exactly match, a nullability warning returns.

**Comment by jnm2** [comment]
The problem goes further than nullability. An error type is getting created by the method type inferrer for the ToDictionary call. This can even be seen in the IDE:

![image](https://github.com/user-attachments/assets/e505c8d0-78ca-480c-8182-e2d9af4c4013)

Versus for exact name matches on the tuple elements:

![image](https://github.com/user-attachments/assets/a1627355-6839-417b-a149-38cc04efc5a9)

There are no diagnostics or emit issues in the final compilation, though.

**Comment by jnm2** [comment]
I looked into this further, and found that the nullability phase might be getting something right. I'm not sure this code should be compiling. If you insert an object cast, the code stops compiling because the tuple element names conflict and are discarded, and `tuple.Name` and `tuple.Value` can't be found:

```cs
#nullable disable
class C
{
    void M((string Name, object Value)[] parameters)
    {
        _ = parameters.Append(("A", (object)"A")).ToDictionary(
            tuple => tuple.Name, // '(string, object)' does not contain a definition for 'Name' [...]
            tuple => tuple.Value); // '(string, object)' does not contain a definition for 'Value' [...]
    }
}
```

Alternatively to adding the object cast, you could change the parameters to `(string Name, string Value)[]`.

So, potentially, I don't think the original example should be compiling at all. The sad part is I know I've taken a dependency on this "only consider names of the receiver's tuple" behavior, over and over, with these LINQ methods.

What's happening is that when the tuple types aren't equal (`(string, object)` versus `(string, string)`), the Append<T> method type inference takes `(string, object)` as an exact bounds for `T` via the `T[]` expression being passed as the receiver parameter, because the spec says that the array element type is an exact bound if the element type is a value type. Whereas the `(string, string)` expression being passed as Append's second parameter becomes a lower bound on Append's T.
Then, when there is an exact bound, only the exact bounds become candidates for T. Then the other bounds are merged or removed.
When it comes time to merge and remove, it all comes down to which way this line evaluates:
<https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs#L3299>
`(string, object)` is not equal to `(string, string)`, so it never calls MergeAndReplaceIfStillCandidate with the lower bound. The lower bound ends up being totally ignored. This means that the conflicting tuple element names are never taken into consideration, and it infers all tuple element names from Append's receiver parameter and ignores the tuple element names from Append's second parameter.

**Comment by jnm2** [comment]
@jcouv Is this connected to <https://github.com/dotnet/roslyn/issues/27961>?

**Comment by jnm2** [comment]
"Inference should be based on unconverted arguments" seems like it sums up the exact discrepancy that this issue was about. The arguments being _passed back in_ to method type inference have already been target-typed using the result of the prior inference for the same type parameter.

![image](https://github.com/user-attachments/assets/ae4d3100-7c87-4f7a-a395-beffdf1777d4)

**Comment by jnm2** [comment]
Perhaps a fix could be to do something like what was done for lambdas and collection expressions in `getArgumentForMethodTypeInference`?

https://github.com/dotnet/roslyn/blob/0cd65d1399a13ef6e0e4ea0ba9575996a957ceea/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L7776-L7797

**Comment by jcouv** [comment]
> Perhaps a fix could be to do something like what was done for lambdas and collection expressions in `getArgumentForMethodTypeInference`?

Yes, that sounds right. The infrastructure for target-typed expressions has evolved over time and I think that collection expressions are the latest example, so may be a good example to follow.
